### PR TITLE
feat: add debug info upload service

### PR DESCRIPTION
**Pull request by korniltsev-grafanista**
## Summary

Add a new `DebuginfoService` that allows profiling agents (e.g., the OTel eBPF profiler) to upload debug information (ELF executables) to Pyroscope's object storage. The symbolizer then uses these uploaded files as an additional source when resolving symbols, falling back to debuginfod if no upload is available.

This is an ALPHA state implementation for internal use and testing. Breaking changes will follow.

### What's included

- **New proto API** (`debuginfo.v1alpha1`): Bidirectional streaming `Upload` RPC using Connect. Clients send a `ShouldInitiateUploadRequest`, receive a response indicating whether to proceed, then stream `UploadChunk` messages. The API is marked `v1alpha1` to signal experimental status.
- **Debug info store** (`pkg/debuginfo/store.go`): Handles upload lifecycle — validates GNU build IDs, checks for duplicate/in-progress/stale uploads via `ObjectMetadata` persisted as JSON in object storage, streams the uploaded binary to the bucket.
- **Upload reader** (`pkg/debuginfo/reader/`): An `io.Reader` adapter that assembles streamed chunks into a contiguous reader for the object storage upload.
- **Symbolizer integration**: The symbolizer now checks the debug info upload bucket (keyed by tenant + GNU build ID) before falling back to debuginfod. The symbolizer bucket was changed from a prefixed bucket to the shared storage bucket, with explicit path prefixes (`symbolizer/`, `debug-info/`).
- **API wiring**: The service is registered on the distributor's HTTP handler with auth and a 5MB per-message limit.

### Key design decisions

- **Own proto, not Parca's**: Inspired by Parca's debug info upload but uses Pyroscope's own protobuf definitions with Connect-Go handlers instead of gRPC.
- **Metadata in object storage**: Upload state (`UPLOADING`/`UPLOADED`) is tracked via a JSON metadata file alongside the binary in the bucket, avoiding the need for an additional database.
- **Stale upload recovery**: If a previous upload started but never finished and exceeds `MaxUploadDuration + 2min`, it's considered stale and can be retried.

## Test plan

- [x] Unit tests for `Store` upload flow, metadata lifecycle, and edge cases (`pkg/debuginfo/store_test.go`)
- [x] Unit tests for `UploadReader` chunk assembly (`pkg/debuginfo/reader/reader_test.go`)
- [x] Symbolizer tests for the uploaded debug info fetch path (`pkg/symbolizer/symbolizer_test.go`)
- [ ] Manual test: upload debug info via profiling agent, verify symbolization uses uploaded file

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Adds a new authenticated streaming upload endpoint that writes executables/metadata to shared object storage and changes symbolization to read from those uploads, which can affect storage usage and symbol resolution behavior if misconfigured or abused.
> 
> **Overview**
> Adds an *alpha* Connect bidirectional streaming `DebuginfoService.Upload` API (`debuginfo.v1alpha1`) plus generated Go/OpenAPI artifacts, enabling agents to upload ELF executables in chunks after an init/"should upload" handshake.
> 
> Implements `pkg/debuginfo.Store` to validate GNU build IDs, track upload state via JSON `ObjectMetadata` in the bucket, detect stale/in-progress uploads, enforce max upload size/duration, and stream chunks into object storage via a new `UploadReader` helper.
> 
> Wires the service into the distributor HTTP API with new `connectOptionsDebugInfo` (auth + 5MB read limit) and configuration flags, and updates the symbolizer to (1) use tenant-scoped object-store keys for cached Lidia data and (2) try fetching uploaded debuginfo (`debug-info/<tenant>/<buildid>/exe`) before calling debuginfod; tests updated/added accordingly.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 68c2faba77c8e38cbb0f48fceffafbaa6627bc72. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: pkg/parca/debuginfo/metadata.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	dbginfo := &debuginfopb.Debuginfo{}
+	if err := protojson.Unmarshal(content, dbginfo); err != nil {
+		return nil, fmt.Errorf("unmarshal debuginfo metadata: %w", err)
+	}
+	return dbginfo, nil
+}
+
+func (m *ObjectStoreMetadata) write(ctx context.Context, dbginfo *debuginfopb.Debuginfo) error {
+	if dbginfo.BuildId == "" {
+		return errors.New("build id is required to wirte debuginfo metadata")
```
Comment (marcsanmi) [inline pkg/parca/debuginfo/metadata.go]:
```suggestion
		return errors.New("build id is required to write debuginfo metadata")
```

File: pkg/parca/debuginfo/store.go [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+		typ      = req.GetInfo().Type
+	)
+
+	ctx := stream.Context()
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(attribute.String("build_id", buildID))
+	span.SetAttributes(attribute.String("upload_id", uploadID))
+
+	if err := s.upload(ctx, buildID, uploadID, typ, r); err != nil {
+		level.Debug(s.logger).Log(
```
Comment (marcsanmi) [inline pkg/parca/debuginfo/store.go]: Should it be `level.Error`?

File: pkg/parca/debuginfo/store.go [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+				return &debuginfopb.ShouldInitiateUploadResponse{
+					ShouldInitiateUpload: true,
+					Reason:               ReasonDebuginfoNotEqual,
+				}, nil
+			//case debuginfopb.DebuginfoUpload_STATE_PURGED:
+			//	// Debuginfo was purged, allow re-uploading
+			//	return &debuginfopb.ShouldInitiateUploadResponse{
+			//		ShouldInitiateUpload: true,
+			//		Reason:               ReasonDebuginfoPurged,
+			//	}, nil
```
Comment (marcsanmi) [inline pkg/parca/debuginfo/store.go]: 🧹 ?

File: pkg/symbolizer/symbolizer.go [hunk 4]
```diff
@@ -484,14 +494,29 @@ func (s *Symbolizer) fetchLidiaFromDebuginfod(ctx context.Context, buildID strin
...
 	if err != nil {
+		s.metrics.debugSymbolResolutionErrors.WithLabelValues(errMetric).Inc()
 		return nil, err
 	}
 
 	return lidiaData, nil
 }
 
+func (s *Symbolizer) fetch(ctx context.Context, buildID string) (io.ReadCloser, error) {
+	if r, err := s.fetchFromParca(ctx, buildID); err == nil {
```
Comment (marcsanmi) [inline pkg/symbolizer/symbolizer.go]: Do you think is worth considering adding a background job to pre-convert uploaded debug info to Lidia format right after upload so the first symbolization request would be faster? Maybe it's just a premature optimization

File: pkg/debuginfo/parcaglue.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+		debuginfov1alpha1grpc.RegisterDebuginfoServiceServer(server, disabled{})
+		return nil
+	}
+	if bucket == nil {
+		return fmt.Errorf("storage bucket is required for debug-info")
+	}
+
+	t := noop.Tracer{}
+	l = log.With(l, "component", "debug-info")
+	bucket = objstore.NewPrefixedBucket(bucket, symbolizer.BucketPrefixParcaDebugInfo)
```
Comment (marcsanmi) [inline pkg/debuginfo/parcaglue.go]: I think we don't have a cleanup mechanism to control unbounded growth (neither for symbolizer path). Could it be an issue?

File: pkg/parca/cache/cache.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+func (c *LRUCache[K, V]) Add(key K, value V) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.lru.Add(key, value)
+}
+
+func (c *LRUCache[K, V]) Get(key K) (V, bool) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+	return c.lru.Get(key)
```
Comment (marcsanmi) [inline pkg/parca/cache/cache.go]: Could it be a race?  I think `.Get` modifies internal state with `MoveToFront` while using a `RLock`

Comment (korniltsev-grafanista) [reply pkg/parca/cache/cache.go]: turns out it's no longer used once I removed the parca's debuginfod logic. Removed the cache package alltogether

File: pkg/symbolizer/symbolizer.go (hunk 4 above)
Comment (korniltsev-grafanista) [reply pkg/symbolizer/symbolizer.go]: I think it's definetly worth and will help with query times, but I don't know how to implement this without separate service / consustent reliable state/queue
https://github.com/grafana/pyroscope-squad/issues/687

File: pkg/debuginfo/parcaglue.go (hunk 5 above)
Comment (korniltsev-grafanista) [reply pkg/debuginfo/parcaglue.go]: I think it's fine to keep lidia's indefinitely. And preferably to remove the debug information files once they are converted to lidia. All of this is hard to implement in a stateless manner in distributors, so I propose to postpone these decision until we have a proper place/way to implement this https://github.com/grafana/pyroscope-squad/issues/687

File: pkg/symbolizer/symbolizer.go (hunk 4 above)
Comment (korniltsev-grafanista) [reply pkg/symbolizer/symbolizer.go]: I've created a separate issue for the cleanup - in case we do this without a separate service

File: pkg/api/api.go [hunk 7]
```diff
@@ -190,6 +190,22 @@ func (a *API) RegisterOverridesExporter(oe *exporter.OverridesExporter) {
...
+		DebuginfoService_MarkUploadFinished_FullMethodName   = "/parca.debuginfo.v1alpha1.DebuginfoService/MarkUploadFinished"
+	)
+	debugInfoGRPCOptions := []RegisterOption{
+		a.WithAuthMiddleware(),
+	}
+	a.RegisterRoute(DebuginfoService_Upload_FullMethodName, d, debugInfoGRPCOptions...)
+	a.RegisterRoute(DebuginfoService_ShouldInitiateUpload_FullMethodName, d, debugInfoGRPCOptions...)
+	a.RegisterRoute(DebuginfoService_InitiateUpload_FullMethodName, d, debugInfoGRPCOptions...)
+	a.RegisterRoute(DebuginfoService_MarkUploadFinished_FullMethodName, d, debugInfoGRPCOptions...)
+}
```
Comment (simonswine) [inline pkg/api/api.go]: I am a bit worried of our mix of GRPC and Connect-Go. None of this really your fault in this PR, but we have currently:

- Cell internal comms which are a mix of GRPC (From Pyroscope v2 project) and connect-go (before v2)
- Public endpoints are all connect-go (accepting GRPC)  or Pyroscope legacy and we do have a endpoint GRPC endpoint for compatibility with OTLP.

So after your PR we would require a client that send both connect-go for Push and GRPC for the uploads. This feels making the above situation worse. 

How do you/others feel about this? Is it worth copying the parca proto into `api/debuginfo` and generate our own handlers clients? 

If we would do that we can reuse the connect handlers and the client would be complete within the api/ package. The endpoints would also work with curl json/grpc/connect and we would be able to reuse e.g. the API docs generation.

Comment (korniltsev-grafanista) [reply pkg/api/api.go]: At first I was against it because I wanted to preserve as much parca code as possible in genuine state, so that it is diffed and updated in case needed easily. I later had to modify quite a lot of code, so now I am actually in favor of 
Yes, let's do this. The only problem is that I've already merged client code =/
I will try to either update the client and backport tomorrow, or disable the new functionality and aim for the next release.

File: lidia/lidia.go [hunk 8]
```diff
@@ -175,7 +176,10 @@ func CreateLidiaFromELF(elfFile *elf.File, output io.WriteSeeker, opts ...Option
 		if symErr != nil {
 			symbols, dynSymErr = elfFile.DynamicSymbols()
 			if dynSymErr != nil {
-				return fmt.Errorf("failed to read symbols from ELF file: %w, %w", symErr, dynSymErr)
+				err := fmt.Errorf("failed to read symbols from ELF file: %w, %w", symErr, dynSymErr)
```
Comment (korniltsev-grafanista) [inline lidia/lidia.go]: this is unrelated, submit separately

**Review by simonswine — APPROVED** [review]
Thanks for working on this.

I think this is good enough for an initial MVP

File: pkg/symbolizer/symbolizer_test.go [hunk 9]
```diff
@@ -30,10 +30,12 @@ type symbolizerInputs struct {
 	Limits   Limits
 }
 
-func newSymbolizerTest(t *testing.T, inp *symbolizerInputs) (*Symbolizer, *mocksymbolizer.MockDebuginfodClient, *mockobjstore.MockBucket) {
```
Comment (simonswine) [inline pkg/symbolizer/symbolizer_test.go:33]: This needs to be rebased before merge. Maybe you just forget to push?

File: pkg/symbolizer/symbolizer.go [hunk 10]
```diff
@@ -484,14 +493,34 @@ func (s *Symbolizer) fetchLidiaFromDebuginfod(ctx context.Context, buildID strin
...
+		return r, nil
+	}
+	return s.fetchFromDebuginfod(ctx, buildID)
+}
+
+func (s *Symbolizer) fetchFromParca(ctx context.Context, buildID string) (io.ReadCloser, error) {
```
Comment (marcsanmi) [inline pkg/symbolizer/symbolizer.go]: maybe (or something similar)?
```suggestion
func (s *Symbolizer) fetchFromUploadedDebugInfo(ctx context.Context, buildID string) (io.ReadCloser, error) {
```

File: pkg/debuginfo/store.go [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+	}
+	if err := s.writeMetadata(ctx, tenantID, id, md); err != nil {
+		return connect.NewError(connect.CodeInternal, fmt.Errorf("failed to write uploading metadata: %w", err))
+	}
+
+	r := debuginforeader.New(ctx, readChunksFromStream(stream))
+	if err := s.bucket.Upload(ctx, ObjectPath(tenantID, id), r); err != nil {
+		return status.Error(codes.Internal, fmt.Errorf("upload debuginfo: %w", err).Error())
+	}
+	md.FinishedAt = timestamppb.New(time.Now())
```
Comment (marcsanmi) [inline pkg/debuginfo/store.go:177]: Should we update the state to `ObjectMetadata_STATE_UPLOADED` around here?

File: pkg/debuginfo/store.go [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+		return chunk.Chunk, nil
+	}
+}
+
+func validateInit(init *debuginfov1.ShouldInitiateUploadRequest) (*ValidGnuBuildID, error) {
+	if init == nil {
+		return nil, fmt.Errorf("first message expected to be init")
+	}
+	if init.File == nil {
+		return nil, fmt.Errorf("init.FileData == nil")
```
Comment (marcsanmi) [inline pkg/debuginfo/store.go]:
```suggestion
		return nil, fmt.Errorf("init.File == nil")
```

File: pkg/debuginfo/store.go [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+		StartedAt:  timestamppb.New(time.Now()),
+		FinishedAt: nil,
+	}
+	if err := s.writeMetadata(ctx, tenantID, id, md); err != nil {
+		return connect.NewError(connect.CodeInternal, fmt.Errorf("failed to write uploading metadata: %w", err))
+	}
+
+	r := debuginforeader.New(ctx, readChunksFromStream(stream))
+	if err := s.bucket.Upload(ctx, ObjectPath(tenantID, id), r); err != nil {
+		return status.Error(codes.Internal, fmt.Errorf("upload debuginfo: %w", err).Error())
```
Comment (marcsanmi) [inline pkg/debuginfo/store.go]: This seems mixed grpc and connect errors.

File: api/debuginfo/v1alpha1/debuginfo.proto [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    //  DEBUGINFO_TYPE_BLAZESYM = 2;
+  }
+
+  string GNU = 1;
+  string Golang = 2;
+  // optional libpf.FileID rom the otel profiler
+  string OpenTelemetry = 3;
+  string Name = 4;
+  DebuginfoType type = 5;
+}
```
Comment (marcsanmi) [inline api/debuginfo/v1alpha1/debuginfo.proto:36]:
```suggestion
message FileMetadata {
    enum Type {
      TYPE_UNSPECIFIED = 0; // several times used as a good practise on enums, feel free to omit it
      TYPE_EXECUTABLE_FULL = 1;
      TYPE_EXECUTABLE_NO_TEXT = 2;
    }
    // GNU build ID.
    string gnu_build_id = 1;
    // Go build ID.
    string go_build_id = 2;
    // OpenTelemetry profiler file ID (libpf.FileID).
    string otel_file_id = 3;
    // Original binary file name.
    string name = 4;
    // Type of the debug info file.
    Type type = 5;
}

```

File: api/debuginfo/v1/debuginfo.proto [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    //  DEBUGINFO_TYPE_BLAZESYM = 2;
+  }
+
+  string GNU = 1;
+  string Golang = 2;
+  // optional libpf.FileID rom the otel profiler
+  string OpenTelemetry = 3;
+  string Name = 4;
+  DebuginfoType type = 5;
+}
```
Comment (korniltsev-grafanista) [reply api/debuginfo/v1/debuginfo.proto]: I am not sure what to do with `TYPE_UNSPECIFIED` ?

**Comment by korniltsev-grafanista** [comment]
Just wanted to clarify, the PR was not ready for review, only the new protobufs and overall intention to use the new protobufs.

I also think we should have an idea how to move forward with infinite data growth (both lidia and debug info)

**Comment by korniltsev-grafanista** [comment]
> I also think we should have an idea how to move forward with infinite data growth (both lidia and debug info)

I talked with @\simonswine and we agreed to solve this later, and merge this for private use, not for public.

**Comment by korniltsev-grafanista** [comment]
@codex review

**Comment by korniltsev-grafanista** [comment]
@codex review

**Comment by korniltsev-grafanista** [comment]
@cursor-bugbot review

**Comment by korniltsev-grafanista** [comment]
@cursor-bugbot review
